### DISCUSSION TITLE
Should `runtime` scoped dependencies be included in main classloader?

### DISCUSSION THREAD

**Discussion by zregvart** [sormuras/junit-platform-maven-plugin#75] [Q&A]
Seems like dependencies scoped as `runtime` are not included in the main classloader. For example if I add`ch.qos.logback:logback-classic` dependency in `runtime` scope, `org.slf4j:slf4j-api` in `compile` scope will not detect it and will issue a warning about missing `org.slf4j.impl.StaticLoggerBinder` class.

The only workaround I've found is to run in isolation of `MERGED` or `NONE`. I've created [an example](https://github.com/zregvart/junit-maven-runtime) to help illustrate this. When running with 1.1.2 I get:

```
$ mvn test
...
[INFO] --- junit-platform-maven-plugin:1.1.2:launch (injected-launch) @ junit-maven-runtime ---
[INFO] Launching JUnit Platform 1.7.2...
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
...
```

And if I change the isolation to `MERGED` or `NONE` I get:

```
$ mvn test
...
[INFO] --- junit-platform-maven-plugin:1.1.2:launch (injected-launch) @ junit-maven-runtime ---
[INFO] Launching JUnit Platform 1.7.2...
12:42:31.608 [pool-3-thread-1] INFO  io.github.zregvart.Log - Hello from main
12:42:31.616 [pool-3-thread-1] INFO  i.g.zregvart.RuntimeClasspathTest - Hello from test
...
```

The real issue I'm having with this is that I would like to declare some dependencies in `runtime` scope so the main code doesn't depend on them at compile time, and have the implementation discovered via service loader mechanism at runtime.

This might be a conscious decision and I'm wondering about the reason behind this?

**Comment by sormuras** [comment]
Hi Zoran,

The underlying reasoning to have separate class loaders is layed out here: https://github.com/sormuras/junit-platform-isolator/issues/1

As I didn't touch boths projects (isolator and this plugin) for months, I'm not sure this behaviour is a bug or a feature. Will try to examine your example later ... have you tried running with the `JAVA` executor? Does that trigger the same results?

**Reply by zregvart** [reply]
Thanks for the reply, running with `<executor>JAVA</executor>` and `<isolation>ABSOLUTE</isolation>` seems to add dependencies in `runtime` scope to the main classloader.

**Reply by zregvart** [reply]
I had a look at the `MavenDriver` and I think to get the main classloader to load both compile and runtime dependencies this line:

https://github.com/sormuras/junit-platform-maven-plugin/blob/9fb4d36e074c9d24a45a3d77e3b4ac642431d0eb/src/main/java/de/sormuras/junit/platform/maven/plugin/MavenDriver.java#L111

Should be:

```java
addAll(project.getRuntimeClasspathElements(), mainPaths); 
```

This will add dependencies in `compile` and in `runtime` scope:

https://github.com/apache/maven/blob/ef62d7a4bd57cdd315dfa187780790e3449321aa/maven-core/src/main/java/org/apache/maven/project/MavenProject.java#L418

One strong argument for including the `runtime` in addition to `compile` scope is within the [Maven documentation](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html#Dependency_Scope):

> runtime
> This scope indicates that the dependency is not required for compilation, but is for execution. Maven includes a dependency with this scope in the runtime and test classpaths, but not the compile classpath.

From my understanding this is what Surefire does, but I struggle to find documentation on this.

**Reply by sormuras** [reply]
Sounds reasonable. Will turn this discussion into an issue...
